src/api/source/TileClient.js

```diff
@@ -0,0 +1,80 @@
+import { validateTemplateURL } from '../url';
+import DataframeCache from './DataframeCache';
+import { rTiles } from '../../client/rsys';
+
+
+export default class TileClient {
+    constructor(templateURLs) {
+        if (!Array.isArray(templateURLs)) {
+            templateURLs = [templateURLs];
+        }
+        templateURLs.map(validateTemplateURL);
+        this._templateURLs = templateURLs;
+        this._requestGroupID = 0;
+        this._oldDataframes = [];
+        this._cache = new DataframeCache();
+    }
+
+    bindLayer(addDataframe, dataLoadedCallback) {
+        this._addDataframe = addDataframe;
+        this._dataLoadedCallback = dataLoadedCallback;
+    }
+
+    requestData(viewport, responseToDataframeTransformer) {
+        const tiles = rTiles(viewport);
+        this._getTiles(tiles, responseToDataframeTransformer);
+    }
+
+    free() {
+        this._cache = new DataframeCache();
+        this._oldDataframes = [];
+    }
+
+    _getTiles(tiles, responseToDataframeTransformer) {
+        this._requestGroupID++;
+        let completedTiles = [];
+        let needToComplete = tiles.length;
+        const requestGroupID = this._requestGroupID;
+        tiles.forEach(t => {
+            const { x, y, z } = t;
+            this._cache.get(`${x},${y},${z}`, () => this._requestDataframe(x, y, z, responseToDataframeTransformer)).then(
+                dataframe => {
+                    if (dataframe.empty) {
+                        needToComplete--;
+                    } else {
+                        completedTiles.push(dataframe);
+                    }
+                    if (completedTiles.length == needToComplete && requestGroupID == this._requestGroupID) {
+                        this._oldDataframes.map(d => d.active = false);
+                        completedTiles.map(d => d.active = true);
+                        this._oldDataframes = completedTiles;
+                        this._dataLoadedCallback();
+                    }
+                });
+        });
+    }
+
+    _getDataframe(x, y, z) {
+        return this._cache.get(`${x},${y},${z}`, () => this._requestDataframe(x, y, z));
```

Comment:
- Calling `_requestDataframe` without the 4º parameter  will make[this](https://github.com/CartoDB/carto-vl/pull/487/files#diff-f44fa8904729341dc09a5c44a14ed8c5R74) fail.

- This function is not used